snuba/cli/devserver.py

```diff
@@ -152,6 +152,19 @@ def devserver(*, bootstrap: bool, workers: bool) -> None:
                     "--consumer-group=metrics_group",
                 ],
             ),
+            (
+                "metrics-aggregates-consumer",
+                [
+                    "snuba",
+                    "multistorage-consumer",
+                    "--storage=metrics_distributions",
+                    "--storage=metrics_counters",
+                    "--storage=metrics_sets",
+                    "--auto-offset-reset=latest",
+                    "--log-level=debug",
+                    "--consumer-group=metrics_aggregates_group",
+                ],
+            ),
```

Comment:
Back to the original comment here. 
If you spin up two independent consumers you will have two independent consumer groups with independent offsets. The raw data and the aggregated ones will not be consistent with each other. 
I think you should jsut add the storages to the existing consumer so they share the same consumer group. 
Also, this way you will be duplicating data in the aggregated table because both the you will be writing both through the aggregate consumer and through the materialzied views that still exist. 

I think there is a larger discussion to decide how to roll this out (and how to roll it back if we decided to go back to mat views) in a consistent way that would not miss or duplicate data.

Could you please come up with a plan (here, no need for more documents) for the steps to achieve this knowing that in production the deployment configuration is not in the snuba code base so config and code cannot be deployed together.

Reply:
We spoke offline, and the `materialization_version` is used by the materialized views to only run if it is equal to 0, so I think I agree that we can combine these consumers into one and change the bucket writing code (in `MetricsBucketProcessor`) to set `materialization_version=1` so that the materialized views and the consumer don't both write to the aggregates tables.

Therefore my deploy plan would be:
1. merge this change (also setting materialization_version to 1 in the buckets processors)
2. ask ops to change the multistorage consumer configuration to also use the direct aggregate table storages as arguments
3. stop the consumer group and record the offset
4. do a deploy to restart the consumer group which now writes direct aggregates

I see two different rollback-requiring situations:
1. the aggregate consumer is too slow and can't keep up with the incoming data
2. the aggregate consumer keeps up with incoming data but corrupts data in one or more of the aggregate tables 

For Situation 1, where VALUES-writing is too slow, we can simply roll back the ops change and then the code change. Since the consumer offset won't advance beyond what the writers can process, reverting to the old logic should just "catch up" over time.

Situation 2 is a bit more complex. To fix data in the aggregate tables, we'll need to "undo" all the writes for materialization_version = 1 and rerun all the writes with materialization_version = 0. The process I'm imagining is:
1. remove all data from aggregate tables where the interval start is after the start time of the first record[s] written with materialization_version = 1
2. roll back the change, setting the materialization_version to 0
3. restart the consumer group from the offset where we stopped using the materialized views. This will cause a large number of row replacements but that should cause the aggregates to be rewritten with the original logic.

Reply:
After step 1 we would be loosing data as the current matview would not be materializing anything and you need to reset the offset to go back. It also would require downtime.
What about this:
1. make a change to this so that the version is taken from settings (not runtime). Also make the aggregated processor rely on the same setting in the opposite way. If the setting says write to matview, write the current version and drop the messages from aggregated processor 
2. change the multistorage config to start the new processors (the aggregated ones). These would do nothing per the setting above.
3. deploy a change that changes the setting. Setting are loaded at startup. So atomically the raw data stop getting materialized (different version) but keeps getting written in the raw. storage and the aggregated storage start writing. 

Rollback strategy: switch the setting back. and deploy.


> remove all data from aggregate tables where the interval start is after the start time of the first record[s] written with materialization_version = 1

This is impossible to achieve consistently as clickhosue keeps merging buckets. Once data is written in the aggregated storage is permanent and cannot be deleted in a consistent way.

Reply:
In that case we're assuming data corruption is not fixable? I can imagine that is OK for preproduction but just want to be sure we're on the same page since theoretically you could recalculate the aggregates from the bucket data.

Reply:
> In that case are we assuming that data corruption is not fixable? I can imagine that is OK for preproduction but just want to be sure we're on the same page since theoretically you could recalculate the aggregates from the bucket data.

I would try very hard to avoid data corruption. When you said theoretically you could build by recalculating the aggregates, were you referring to dropping the whole materialized view and creating it again from scratch? Even if you did that you would have to stop the Kafka consumer till you delete/create the materialized view on all nodes of the cluster. Creating it from scratch would take a lot of time I belive.

Reply:
Expanding on data corruption. What kind of corruption are you talking about ?
- data is written in a way that clickhouse cannot read. This is a disaster recovery scenario and generally the outcome of a clickhouse bug otherwise it would not allow you to write, depending what happened you may have to restore from backup. You cannot just restore from raw data as the raw data will always have lower retention that aggregated data. If you cannot read parts because they are corrupted on disk, clickhouse does not allow you to access the table entirely. More likely it will exclude broken parts.
- We wrote data we should not have written by mistake. In this case you cannot really redress the result consistently (but it is the same kind of issues in any aggregated store). Once clickhouse merged a bucket (imagine a daily bucket) that contains good and bad data you cannot separate them unless you throw away the entire bucket. 
- We did not write stuff we should have written. In this case you could materialize data from the raw table.

All these scenarios are disaster recovery scenarios and would happen with the database not accessible for the whole operation.

So the goal of being able to revert is basically something we would do in pre-production and just in case keep around in case we start seeing errors later. Imagine INSERTs failing for some reasons. 
The reason for doing it with no downtime is because it makes it a lot easier operationally but also in terms of implementation as you do not have to build anything to deal with offset based synchronization which we do not have and simply. .... the procedure is simpler: switch a setting.

Reply:
> * We wrote data we should not have written by mistake. In this case you cannot really redress the result consistently (but it is the same kind of issues in any aggregated store). Once clickhouse merged a bucket (imagine a daily bucket) that contains good and bad data you cannot separate them unless you throw away the entire bucket.

I was thinking of this situation, and specifically a subset case where we wrote the correct data to the bucket table, but we're misunderstanding something about the aggregate merge tree, so only the aggregate table has some kind of bad data. In such a situation, I can imagine in theory using the bucket table to recreate *only* bad data in the aggregate table, but I also believe you when you say that this is not feasible in practice.

I am convinced of using a setting to control which way we write to the aggregate tables -- I might ask, though, why not have it be a runtime config so that we can turn it on/off ourselves without involving ops?

Reply:
runtime config, being runtime, is not going to be synchronized between processors. 
If each processor (buckets and aggregated) independently load the runtime config, for the same message one could read 0 and the second may read 1 if the value changed in between. So you may end up double writing a batch if both processors think they have to write or not writing it at all if none thinks they need to write.
settings are read at startup, the entire consumer has the same value and it is only re-read upon restart.

Reply:
Sounds good, will add this to settings and follow your proposed approach

Reply:
I've implemented the changes you suggested.